src/typecheck.rs

```diff
@@ -1,42 +1,215 @@
-use crate::*;
-use indexmap::map::Entry as IEntry;
+use std::ops::AddAssign;
+
+use crate::{constraint::AllEqualTypeConstraint, *};
+use hashbrown::HashMap;
 use typechecking::TypeError;
 
+#[derive(Clone, Debug, PartialEq, Eq, Hash, Default)]
+pub struct Actions(pub(crate) Vec<NormAction>);
+impl Actions {
+    fn subst(&mut self, subst: &HashMap<Symbol, AtomTerm>) {
+        let actions = subst.iter().map(|(symbol, atom_term)| match atom_term {
+            AtomTerm::Var(v) => NormAction::LetVar(*symbol, *v),
+            AtomTerm::Literal(lit) => NormAction::LetLit(*symbol, lit.clone()),
+            AtomTerm::Global(v) => NormAction::LetVar(*symbol, *v),
+        });
+        let existing_actions = std::mem::take(&mut self.0);
+        self.0 = actions.chain(existing_actions).collect();
+    }
+}
+
+// TODO: implement custom debug
+#[derive(Debug, Clone)]
+pub struct UnresolvedCoreRule {
+    pub body: Query<Symbol>,
+    pub head: Actions,
+}
+
+#[derive(Debug, Clone, Default)]
+pub struct ResolvedCoreRule {
+    pub body: Query<ResolvedSymbol>,
+    pub head: Actions,
+}
+
+#[derive(Debug, Clone)]
+pub enum ResolvedSymbol {
+    Func(Symbol),
+    Primitive(Primitive),
```

Comment:
I'm still of the opinion that equality is special and should not be encoded as the `value-eq` primitive. It doesn't make sense because we never actually call `value-eq`, we compile it into equality checks. This will be confusing for new people in the code base.

Reply:
I am convinced. After thinking about it, I introduced another rule variant and the current core definition looks like this:

```
#[derive(Debug, Clone)]
pub struct UnresolvedCoreRule {
    pub body: Query<SymbolOrEq>,
    pub head: Actions,
}

#[derive(Debug, Clone)]
pub struct CanonicalizedCoreRule {
    pub body: Query<Symbol>,
    pub head: Actions,
}
```

And `canonicalize` transforms an `UnresolvedCoreRule` into a `CanonicalizedCoreRule`. This makes equalities explicit in our IR definition.

Reply:
Great!